web/js/codeworld-mode.js

```diff
@@ -82,43 +116,91 @@ CodeMirror.defineMode('codeworld', (_config, modeConfig) => {
 
         if (stream.match(RE_DASHES)) {
             stream.skipToEnd();
-            return 'comment';
+            return ['comment'];
         }
 
-        if (stream.match(RE_QUAL)) return 'qualifier';
-        if (stream.match(RE_VARID) || stream.match(RE_VARSYM)) return 'variable';
-        if (stream.match(RE_CONID) || stream.match(RE_CONSYM)) return 'variable-2';
-        if (stream.match(RE_NUMBER)) return 'number';
-        if (stream.match(RE_CHAR) || stream.match(RE_STRING)) return 'string';
+        if (checkContextStarter(stream, state)) {
+            return ['keyword'];
+        }
+
+        if (state.scanState === NORMAL && stream.match(RE_IN)) {
+            // Cancel previous layout 'let' context
+            state.layoutContext.pop();
+            return ['keyword'];
+        } else if (state.scanState === NORMAL) {
+            let lastContext = state.layoutContext[state.layoutContext.length - 1];
+            lastContext = lastContext === '{' ? 0 : lastContext;
+            if (stream.sol() && stream.indentation() < lastContext) {
+                // Close previous context
+                state.layoutContext.pop();
+            }
+            if (stream.column() === lastContext || stream.column() === 0) {
+                spanStyles.push('layout');
+            }
+        } else if (state.scanState === TRIGGERED && stream.match('{')) {
+            // Cancel previous layout context because of next char is '{'
+            state.layoutContext.pop();
+            state.scanState = NORMAL;
+        } else if (state.scanState === TRIGGERED) {
+            state.layoutContext.push(stream.column());
+            spanStyles.push('layout');
+            state.scanState = NORMAL;
```

Comment:
What about this?

    let
      let x = 4
          y = 6
      z = 7
    in x + y + z

It looks like you're missing a layout context here.  This logic needs to come earlier.

Reply:
I not got this example, sorry
![image](https://user-images.githubusercontent.com/5370736/54475552-7ebe1900-4803-11e9-8668-2134cff35765.png)

Reply:
My mistake.  I gave you a bad example.  Another one might be:

    do
        do one
           two
        three

The second `do` starts a new layout context, but also triggers nested layout.  On further though, I think this is only possible with `do`, and granted I don't really use `do` in codeworld-mode.  It's still possible, though, and not hard to fix.